### PR TITLE
Integrating Meilisearch Indexing with Directus

### LINKED ISSUE #220: Integrating Meilisearch Indexing in SolidStart and Directus
### What is your idea?

- Installing SolidStart and setting up a new project.
- Create a Directus Helper
- Creating Collections
- Installing and setting up Meilisearch
- Integrating Directus with Meilisearch
    - Writing Directus hooks to send data to Meilisearch
         - On data creation
         - On data update
         - On data deletion
     
- Automating Data Synchronization
    - Using Directus webhooks to automate data indexing in Meilisearch
    - Handling real-time updates and ensuring data consistency.  
    
- Performing Search Queries
    - Writing search queries using Meilisearch API
    - Integrating search functionality into Directus
    - Displaying search results in the Directus interface
    
- Handling Advanced Search Features
   - Implementing filtering and sorting options
   - Utilizing Meilisearch advanced features (faceted search, typo tolerance)

### What are the key takeaways from your post?

By the end of this tutorial, the reader will learn:
- Setting Up the Environment: Install and configure both Directus and Meilisearch to ensure they are running properly.
- Creating a Directus Collection: Create and define a new collection in Directus, populating it with sample data.
- Configuring Meilisearch: Create and configure an index in Meilisearch, including adding sample data.
- Integrating Directus with Meilisearch: Set up API endpoints in Directus and write scripts or use Directus hooks to - synchronize data between Directus and Meilisearch during creation, updates, and deletion.
- Automating Data Synchronization: Use Directus webhooks to automate the indexing process in Meilisearch and ensure real-time updates and data consistency.

### Country of residence

Nigeria

### Terms & Conditions

- [X] You understand that submitting this form does not mean we accept your proposal. We will accept proposals at our discretion.
- [X] You understand that to be paid you will need to generate an invoice with a 30 day payment term and send it to us once the final draft of your work is accepted (we will share how to do this once finished).
- [X] You understand that you will require a bank account in one of the countries listed here (https://help.bill.com/direct/s/article/360007172671).
- [X] You understand that you need to set up a Directus project by either running locally, self-hosting, or using Directus Cloud, but being a guest author does not entitle you to extended trials or free credit.
- [X] You understand that you will grant us an unlimited license to post and promote your work as part of this program. You may repost, but using our copy as the canonical source.
- [X] You are aged 18 or over.

### PR DESCRIPTION
**Pull request by Claradev32**
This is the PR for the issue https://github.com/directus-labs/guest-authoring/issues/220

### CODE REVIEW

**Review by phazonoverload — COMMENTED** [review]
I'm looking to review this imminently, but could you please remove the SolidStart pieces as we agreed in the issue? I'm only interested in maintaining the index from Directus.

**Comment by Claradev32** [comment]
Hi @phazonoverload 
I have updated the draft to remove the SolidStart part. 

Thank you

File: integrating-melisearch-indexing-in-solidStart-and-directus/index.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+```javascript
+import { MeiliSearch } from 'meilisearch'
+
+const client = new MeiliSearch({
+  host: process.env.MEILISEARCH_HOST,
+  apiKey: process.env.MEILISEARCH_API_KEY
+})
+const index = client.index('articles')
+```
```
Comment (phazonoverload) [inline integrating-melisearch-indexing-in-solidStart-and-directus/index.md:65]:
````suggestion
## Setting Up Meilisearch

Sign up for a Meilisearch account if you haven't already. Once you have your Meilisearch instance details, you will be able to copy your credentials in your dashboard. 

![Melisearch dashboard](<Screenshot 2024-07-11 at 06.45.38.png>)

Add the following environment variables to your project: 

```docker-compose
MEILISEARCH_HOST=your_meilisearch_host
MEILISEARCH_API_KEY=your_meilisearch_api_key
```

Navigate into your new extension directory, run `npm install meilisearch`, and then `npm run dev` to start the automatic extension building.

At the top of your extension's `src/index.js` file, initialize the Meilisearch client:

```javascript
import { MeiliSearch } from 'meilisearch'

const client = new MeiliSearch({
  host: process.env.MEILISEARCH_HOST,
  apiKey: process.env.MEILISEARCH_API_KEY
})
const index = client.index('directus_index')
```
````

File: integrating-melisearch-indexing-in-solidStart-and-directus/index.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+Update your extension's exported function to process create events when a new `article` is added to the collection:
+
+```javascript
+export default ({ action }) => {
+  action('articles.items.create', async (meta) => {
+    await index.addDocuments([{ id: meta.key, ...meta.payload }])
+  })
+}
+```
+The `articles.items.create` action hook triggers after item creation. The `meta` object contains the new item's key (ID) and other fields in its payload property. By setting the `objectID` to the Directus item id, we ensure accurate referencing and management in Meilisearch.
```
Comment (phazonoverload) [inline integrating-melisearch-indexing-in-solidStart-and-directus/index.md:80]:
````suggestion
## Saving New Items to Index

Update your extension's exported function to process create events when a new `article` is added to the collection:

```javascript
export default ({ action }) => {
  action('articles.items.create', async (meta) => {
    await index.addDocuments([{ id: meta.key, ...meta.payload }])
  })
}
```

The `articles.items.create` action hook triggers after item creation. The `meta` object contains the new item's `key` (ID) and other fields in its `payload` property. By setting the `objectID` to the Directus item `id`, we ensure accurate referencing and management in Meilisearch.
````

File: integrating-melisearch-indexing-in-solidStart-and-directus/index.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  action('articles.items.update', async (meta) => {
+    await Promise.all(
+      meta.keys.map(async (key) => 
+        await index.updateDocuments([{ id: key, ...meta.payload }])
+      )
+    )
+  })
+}
+```
+The `articles.items.update` action hook triggers when articles are updated. It receives `meta.keys` (an array of updated item IDs) and `meta.payload` (changed values). The hook updates each modified document in Meilisearch using `Promise.all` for efficient processing.
```
Comment (phazonoverload) [inline integrating-melisearch-indexing-in-solidStart-and-directus/index.md:96]:
````suggestion
### Updating Items in Index

Add another action hook to process updates when one or more articles are modified:

```javascript
action('articles.items.update', async (meta) => {
  await Promise.all(
    meta.keys.map(async (key) => 
      await index.updateDocuments([{ id: key, ...meta.payload }])
    )
  )
})
```

The `articles.items.update` action hook triggers when articles are updated. It receives `meta.keys` (an array of updated item IDs) and `meta.payload` (changed values). The hook updates each document in Meilisearch.
````

File: integrating-melisearch-indexing-in-solidStart-and-directus/index.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+   action('articles.items.delete', async (meta) => {
+    await index.deleteDocuments(meta.keys)
+  })
+}
+```
+The `articles.items.delete` action hook triggers when articles are deleted. It receives `meta.keys`, an array of deleted item IDs. The hook uses these keys to remove the corresponding documents from the Meilisearch index.
+
+Now add 3 items to your articles collection.
+
+![Melisearch with data from Directus](<Screenshot 2024-07-11 at 06.58.04.png>)
```
Comment (phazonoverload) [inline integrating-melisearch-indexing-in-solidStart-and-directus/index.md:114]:
````suggestion
### Deleting Items in Index

Add an action hook to remove items from Meilisearch when they're deleted in Directus:

```javascript
action('articles.items.delete', async (meta) => {
  await index.deleteDocuments(meta.keys)
})
```

The `articles.items.delete` action hook triggers when articles are deleted. It receives `meta.keys`, an array of deleted item IDs. The hook uses these keys to remove the corresponding documents from the Meilisearch index.

Now add 3 items to your articles collection and you should see them in your Meilisearch index.

![Melisearch with data from Directus](<Screenshot 2024-07-11 at 06.58.04.png>)
````

File: integrating-melisearch-indexing-in-solidStart-and-directus/index.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+
+## Summary
+In this tutorial, you've learned how to integrate Meilisearch indexing with Directus and SolidStart. You've learned how to setup the Directus hooks that automatically indexes data created, updated, or deleted from a Directus project in Meilisearch.
```
Comment (phazonoverload) [inline integrating-melisearch-indexing-in-solidStart-and-directus/index.md:118]:
```suggestion
In this tutorial, you've learned how to integrate Meilisearch with Directus. You've learned how to setup the Directus hooks that automatically indexes data created, updated, or deleted from a Directus project in Meilisearch.
```

**Review by phazonoverload — CHANGES_REQUESTED** [review]
Most suggestions are to bring this in line with the format of our existing Algolia post more closely. Once you accept them, we can accept this post 😄

**Comment by Claradev32** [comment]
Alright, I have done that.

Thank you for the suggestions.

**Comment by phazonoverload** [comment]
Thank you for writing this, and we are happy to consider it done in it's current state. My team will schedule it in to be published. 

This repo is just for authoring/reviewing - the actual piece will be published via a Directus project, so we'll take it from here. I'll close this issue to keep the main branch clean, but this is a acceptance of your work.

Could you email us at [devrel@directus.io](mailto:devrel@directus.io) and we'll share what we need to get you paid. Thanks!

**Pull request closed without merging**